DESIGN.md

```diff
@@ -0,0 +1,200 @@
+
+### Indexing Packages
+
+#### Open-source, npm packages first
+
+The catalog will initially be oriented around open-source packages published to npm. This is where the vast majority of elements are now published.
+
+Cataloging publically available closed-source elements (those sold to the public for a fee) is not a goal, as that would require indexing proprietary information about where to retrieve elements, their license and cost, etc.
+
+Package sources other than npm, like GitHub packages or package-less distributions available via HTTP, or closed-source internal elements, may be included in the future. npm sources will be namespaced to prevserve the ability to add new sources.
+
+#### Permissionless importing
+
+We want to make importing packages as transparent as possible, so that they require no special permission or ownership of packages to import. Any anonymous users should be able to request that an npm package is imported. So we need to make importing fast, idempotent (when there is nothig to update), and rate-limited to protect against abuse and DOS.
+
+The catalog will record in the database when a package was imported, what the versions were at import time, and whether the package was a valid, custom-elements containing packge. This will let us know quickly and cheaply whether to read any information from npm to attempt another import.
+
+Packages contain a number of versions and versions are immutable, so if we see that a new version has not been published, we don't need to do a new import.
+
+Once a package has been successfully imported at least once, it will be periodically scanned for new versions to be imported (or we will listen for new versions from the npm registry if possible).
+
+#### Collections
+
+Design systems and other groups of components benefit from first-class support in the catalog. Users may want to search for collections of components rather than individual components.
+
+Collections don't neccesaarily correspond to packages. Many design systems publish each component as a separate npm package.
+
+We would like a way for element authors to define their own collections controll exactly what elements are included. For this it would be ideal to have a centralized definition of a collection. Even though a collection is not published to npm, a collection definition could be published to a well-known URL, such as a file stored in GitHub. A custom element manifest can then include references to collection definitions so that while indexing element we can discover and index collections.
+
+### Schema
+
+The catalog server has two schemas in play: A GraphQL schema and an implcit Firestore schema (implicit, since Firestore is schemaless). We will try to keep these as close as possible. Some places where schemas differ:
```

Comment:
```suggestion
The catalog server has two schemas in play: A GraphQL schema and an implicit Firestore schema (implicit, since Firestore is schemaless). We will try to keep these as close as possible. Some places where schemas differ:
```